File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    visit = [False] * len(tickets)
    
    def dfs(path):
        if len(path) == len(tickets) + 1:
            answer.append(",".join(path))
            return
        
        start = path[-1]
        
        for index, ticket in enumerate(tickets):
            if ticket[0] == start and not visit[index]:
                visit[index] = True
                path.append(ticket[1])
                dfs(path)
                path.pop()
                visit[index] = False
        
    dfs(["ICN"])
    
    return sorted(answer)[0].split(",")

File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    const answer = [];
    const visit = new Array(tickets.length).fill(false);
    
    const dfs = (path) => {
        if (path.length === tickets.length + 1) {
            answer.push(path.join());
            return;
        }
        
        const start = path.at(-1);
        
        tickets.forEach((ticket, index) => {
            if (ticket[0] === start && !visit[index]) {
                visit[index] = true;
                path.push(ticket[1]);
                dfs(path);
                path.pop();
                visit[index] = false;
            }
        });
    };
    
    dfs(["ICN"]);
    
    return answer.sort()[0].split(",");
}
